### ISSUE TITLE
npm@5 prerelease tracking

### ISSUE THREAD

**Issue by zkat** [npm/npm#16510]
This is the official npm@next tracking issue listing known issues and, if relevant, workarounds. The beta ran until May 25th, and we're doing a bit more during the prerelease, to hopefully get `npm@5` into `latest` by May 30th, in time for `node@8`'s release.

You no longer need to install the npm5 package to try this: just do `npm i -g npm@next`.

For the general changelog and release-branch tracking issue, see #16244. The feature list for npm5 is listed there.

Want to help? Check for [tagged `npm5` issues](https://github.com/npm/npm/issues?q=is%3Aopen+is%3Aissue+label%3Anpm5), or pick up any of the open bugs in [`pacote`](https://github.com/zkat/pacote/issues), [`cacache`](https://github.com/zkat/cacache), or [`make-fetch-happen`](https://github.com/zkat/make-fetch-happen). These libraries are all fundamental infrastructure for the new cache, and any improvements will have a direct effect on npm5 itself.

### Known Issues

The following issues are known, and will (hopefully!) be fixed by the time npm5 goes to `latest` on May 30th:

* The new `notice` loglevel messes with some tools that use `npmlog`. For example, you might see verbose log output from `node-pre-gyp` during lifecycle script execution.

* cross-platform `package-lock.json` is still pending some WIP patches. _Soooooooon_

* Some installs using github enterprise with 2FA and `git+ssh` are failing to authenticate. (#16726)

* Global installations fail sometimes.

### Fixed Issues

* **FIXED** in`beta.34` ~a bad commit accidentally broke cache corruption recovery. Oops! Fix forthcoming.~

* **FIXED** in `beta.35` ~dependencies in package.json with spaces in them fail to match (#16520)~

* **FIXED** in `beta.36` ~dependencies with `=` don't resolve (#16524)~

* **FIXED** in `beta.37` (thanks, @zarenner!) ~npm5 fails to authenticate for registries that aren't hosted at the host root (#16528)~

* **FIXED** in `beta.38` This wasn't an npm5 bug per se, but some packages had tarballs with package.json data that did not match the package metadata uploaded (???). npm5 compensates a bit for this now. ~There's apparently some situations where extraneous packages get added, or correct packages get marked as extraneous (#16517)~

* **FIXED** in `beta.43` (thanks, @ivanph!) ~npm5 sends original auth headers with tarballs hosted outside of the manifest registry (https://github.com/zkat/pacote/issues/87, https://github.com/npm/node-fetch-npm/issues/1)~

* **FIXED** in `beta.46` ~Lockfiles are still generated off the filesystem, so they are not yet fully cross-platform.~

* **FIXED** in `beta.46` ~Bundled dependencies might be replaced even if marked as bundled in your `package-lock.json`. This should only yield a warning, since npm will simply replace acceptable versions of those dependencies with regular downloads.~

* **FIXED** in `beta.46` ~Local file/directory dependencies are not complete and will have weird interactions with the `package-lock.json`. This is pending merging https://github.com/npm/npm/tree/link-specifier (#16536)~

* **FIXED** in `beta.47` ~Global installs are failing with a `cannot read 'package' of null` error (#16603)~

* **FIXED** in `beta.48` ~npm crashes with a url not being a valid semver spec (#16600)~

* **FIXED** in `beta.50` ~npm fails to send auth data to scoped private registries~

* **FIXED** in `beta.56` ~`package.json` might be saved even if `package-lock.json` fails to save for some reason. (nb: this should happen much less as of `beta.46`)~

* **FIXED** in `beta.56` ~The installer is still erroneously marking some packages as extraneous (#16517)~

* **FIXED** in `beta.57` ~`npm update` will by default save any updated `devDeps` as if they were `dependencies`.~

* **FIXED** in `beta.58` (Thanks, @nlkluth!) ~Downloads the redirect from http to https and vice versa are not supported right now (#16613, https://github.com/zkat/make-fetch-happen/issues/26)~

* **FIXED** in `beta.59` and `beta.60` (Thanks, @yyjdelete!) ~http->https tunneling proxies don't work~

* **FIXED** in `beta.61` ~`uid must be an unsigned int` error during install (#16544)~

* **FIXED** in `beta.63` ~Using a proxy will keep the proxy agent alive, preventing the node process from exiting. You can safely `^C` out if this happens.~

* **FIXED** in `beta.63` ~Some users have reported symlink issues while upgrading the beta. We tracked down the race, and there should be a fix very soon (#16686)~

* **FIXED** in `beta.64` ~Installing with `sudo` will write cache entries as `root` instead of dropping permissions for cache writes (#16707)~

* **FIXED** in `beta.69` ~username/password registry auth doesn't seem to be working at least for some people~

* **FIXED** in `beta.69` ~package-lock entries previously marked as `optional: true` or `dev: true`, or even `bundled: true` are having those settings clobbered on the following install and written back as full, regular deps.~

* **FIXED** in `npm@5.0.0` ~The `prepare`, `prepublish`, `prepublishOnly` dance isn't quite set in stone yet. Things might be a bit confusing. (https://github.com/npm/npm/issues/16685)~

* **FIXED** in `npm@5.0.0` ~directory dependencies aren't quite done yet in a couple of cases: you might see weird issues with dependencies of dir dependencies causing errors, or even `Max callstack exceeded` errors for certain combinations.~

**Comment by boennemann** [comment]
Congratulations on the release 🎉 ✨

Just saw the man entry is missing for the npm5 binary. I think that's caused by absolute paths in the registry metadata, but I'm not sure b/c npm4 has absolute paths, too, but that works.

**Comment by montogeek** [comment]
I am getting this error when using a local package in npm 5.0.0-beta.34
```
npm WARN saveError Problems were encountered
npm WARN saveError Please correct and try again.
npm WARN saveError invalid: have e2-scripts-utils@1.0.0 (expected: file:../../lib/utils) /Users/montogeek/project/node_modules/e2-scripts-utils
npm WARN saveError extraneous: e2-scripts-utils@1.0.0 /Users/montogeek/project/node_modules/e2-scripts-utils
npm WARN saveError extraneous: jump.js@1.0.2 /Users/montogeek/project/node_modules/jump.js
```

package.json
```
"e2-scripts-utils": "file:../../lib/utils"
```
It works without problem when using npm 4.1.2, executing all lifehooks in the dependency

**Comment by zkat** [comment]
@boennemann we have general issues installing the manpages (turns out they never _quite_ worked, but that's a separate thing to explain). Getting manpages to work again just for the beta is far down the list. Thanks for letting me know, though!

And thank you. I'm excited for people to try this out and help make it better ^_^

**Comment by simonua** [comment]
@zkat, congratulations on the big milestone! I bet that's a load off your shoulders for a quick second. =)

I was wondering whether you or anyone has tried credentialed access to private npm or third party registries such as Microsoft's Package Management. We are a user of the latter and trying beta 36 just now I am getting an E401 - Unauthorized error for all package requests. I'll be glad to make this a side "issue" to talk through to get it out of this thread if that's helpful. Would have to scrub logs a bit, perhaps, if I were to post any. Pinging Microsoft for awareness: @zarenner

**Comment by zkat** [comment]
@simonua Please make a new issue with a debug log. This is a thing I was really hoping to test out. The auth code was completely rewritten, as was all the proxy stuff.

**Comment by simonua** [comment]
@zkat, 10-4

**Comment by zkat** [comment]
@simonua hey! beta.37 is out with a hopefully-fixed auth bit. Wanna give it another shot? Thanks @zarenner for the speedy fix over in https://github.com/zkat/pacote/pull/85 💃

**Comment by simonua** [comment]
Definitely, @zkat! Might be tomorrow morning, but I will certainly report my findings. Great stuff and thank you and @zarenner!

**Comment by zarenner** [comment]
I just opened https://github.com/zkat/pacote/issues/87 which is another issue which affects registries that redirect to another host where Authorization must not be forwarded. So (at least for VSTS specifically) auth issues aren't yet completely resolved.

**Comment by zkat** [comment]
just dropping a note: we're going to have a brief lull in fixes/patches for a couple of days while we work on some fixes for the bigger issues folks are having, and I'm also working on getting a beta into node 8. Shiny new beta releases will ramp back up once that stuff lands :)

Thanks all for the time you've taken to test stuff, and the excellent tickets y'all have filed!

**Comment by zkat** [comment]
Additionally, we've added a TODO list to the bottom of the OP in #16244.

**Comment by zkat** [comment]
k, we're back in business. `package-lock.json` should now be cross-platform, idempotent, etc. 👍

**Comment by montogeek** [comment]
In `beta.46` I am getting:
```
npm ERR! Maximum call stack size exceeded
```

for the same exact `package.json` file

**Comment by christopherstott** [comment]
I just opened an issue - for a change to the lockfile that would be really helpful for CI services - https://github.com/npm/npm/issues/16580

**Comment by zkat** [comment]
@montogeek hah, nice.

Do you have a package.json you can share that reproduces this?

**Comment by montogeek** [comment]
@zkat https://gist.github.com/montogeek/05280165827ca92491de1fe29b6b80f3
There is a local dependency that unfortunately I can't share

**Comment by legodude17** [comment]
From #16601:
> When attempting to install a package locally, npm5 throws ENOENT -4058.

**Comment by Timer** [comment]
> directory dependencies aren't quite done yet in a couple of cases: you might see weird issues with dependencies of dir dependencies causing errors

In the below error, `/Users/joe/Documents/Development/OSS/create-react-app/packages/react-scripts/node_modules/tsconfig-react-app` is a symlink installed by lerna. I'm not sure if that falls under the above quote, but I figured I'd post here to be safe. 😄

```
Joes-MacBook-Pro:react-scripts joe$ npm5 i
npm notice BETA npm5 is beta software. Please see https://github.com/npm/npm/issues/16510 for known issues, and please file a new one if you find something new.
npm WARN checkPermissions Missing write access to /Users/joe/Documents/Development/OSS/create-react-app/packages/react-scripts/node_modules/tsconfig-react-app
npm ERR! path /Users/joe/Documents/Development/OSS/create-react-app/packages/react-scripts/node_modules/tsconfig-react-app
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall access
npm ERR! enoent ENOENT: no such file or directory, access '/Users/joe/Documents/Development/OSS/create-react-app/packages/react-scripts/node_modules/tsconfig-react-app'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/joe/.npm/_logs/2017-05-22T01_07_47_345Z-debug.log
```

**Comment by igalic** [comment]
I'm having trouble self-updating `npm5`:

```
 ~ % sudo -H npm5 i -g npm5
npm notice BETA npm5 is beta software. Please see https://github.com/npm/npm/issues/16510 for known issues, and please file a new one if you find something new.
npm ERR! uid must be an unsigned int

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-05-22T08_02_18_642Z-debug.log
 ǃ > ~ %
```

where `/root/.npm/_logs/2017-05-22T08_02_18_642Z-debug.log` contains the following:

```
33 silly install executeActions
34 silly doSerial global-install 9
35 verbose correctMkdir /root/.npm/_locks correctMkdir not in flight; initializing
36 verbose lock using /root/.npm/_locks/staging-a072192f34a17023.lock for /usr/lib/node_modules/.staging
37 silly doParallel extract 9
38 silly extract npm5@5.0.0-beta.57
39 silly pacote trying npm5@https://registry.npmjs.org/npm5/-/npm5-5.0.0-beta.57.tgz by hash: sha512-RVPYoZIj7+wYFlMbesm3vugJtiqdlpvLUfiKyLZbIOlXY0i7kqgBR8qmTk1yGCj43R1EyzJggcnWDoCYlgZkwA==
40 verbose stack TypeError: uid must be an unsigned int
40 verbose stack     at TypeError (native)
40 verbose stack     at Object.fs.chown (fs.js:1204:11)
40 verbose stack     at Object.chown (/usr/lib/node_modules/npm5/node_modules/graceful-fs/polyfills.js:259:17)
40 verbose stack     at /usr/lib/node_modules/npm5/node_modules/chownr/chownr.js:12:43
40 verbose stack     at go$readdir$cb (/usr/lib/node_modules/npm5/node_modules/graceful-fs/graceful-fs.js:149:14)
40 verbose stack     at FSReqWrap.oncomplete (fs.js:123:15)
41 verbose cwd /home/igalic
42 verbose Linux 4.10.0-21-generic
43 verbose argv "/usr/bin/nodejs" "/usr/bin/npm5" "i" "-g" "npm5"
44 verbose node v6.10.3
45 verbose npm  v5.0.0-beta.56
46 error uid must be an unsigned int
47 verbose exit [ 1, true ]
```

**Comment by simonua** [comment]
@igalic, take a look at #16544 as that looks similar please.

**Comment by rmg** [comment]
npm5 (5.0.0-beta.61) executes the `preinstall` lifecycle script before extracting the package in to the `.staging` directory but `npm@<5` run `preinstall` after the package contents have been extracted.

Not sure if this is one of the expected breaking changes or not.

@zkat do you want things like this opened as separate issues?

**Comment by zkat** [comment]
@rmg go ahead and put up a separate issue! This is worth looking into. We've been moving around a lot of this stuff. Thanks a bunch 🎉

**Comment by corbinu** [comment]
@zkat When are the 5 betas going to be available on the regular npm package? Was trying to update the node 8.0.0 RC.1 from npm beta 56 to 61. Or should I just still install npm5 and use that instead?

**Comment by zkat** [comment]
@corbinu npm5 is slated to hit npm@next tomorrow (probably late in the day)

**Comment by corbinu** [comment]
Great thanks!

**Comment by Fishrock123** [comment]
@zkat I've left a comment in that `preinstall` issue thread (https://github.com/npm/npm/issues/16706#issuecomment-304038178), that breaks [`llnode`](https://www.npmjs.com/package/llnode)'s install right now with npm5.

**Comment by simonua** [comment]
@zkat and team, congratulations on the 5.0.0 release! Awesome stuff! 

Will you maintain the 5.0.0-beta.x releases if or while remaining items on the npm5 bug tracker may be worked through between now and Tuesday? I want to keep working with npm5 side-by-side, but I am not quite comfortable yet replacing 4.6.1 as my rock-solid backup.

Thank you! =)

**Comment by janl** [comment]
@zkat et.al. congrats on the release!

One question: we have the need for a `postinstall` script to differentiate between `npm install` in the source directory of our `module` and `npm install --save module` — pre npm5, we’ve used `process.env.npm_config_save` as a good enough approximation to get the behaviour we need, but with `--save` becoming default in npm5, `npm_config_save` is true for `npm install` in the `module` source directory.

It could be argued, that even with `--save` being default in npm5, that `npm_config_save` shouldn’t be true, when `npm install` is run without a module name specified, but I could imagine that this is problematic for installation.

Is there another way, or a recommended way of differentiating between `npm install` in the `module` source directly and `npm install `module`?

**Comment by rmg** [comment]
@janl is parsing `process.env.npm_config_argv` an option?

**Comment by janl** [comment]
@rmg that was my first idea too, but that's just the command of the `postinstall` script.

**Comment by rmg** [comment]
@janl are you sure? Looks like it's the top-level/initiating CLI args in my testing.. Unless that's another v5 change.

**Comment by iarna** [comment]
@rmg I've gone down the parsing the `config_argv` route before with `in-publish` and it's limited and brittle. Honestly, if you need to distinguish the two modes I think I'd prefer that we have different lifecycles.

@janl What are you doing in your `preinstall` for a published module that you want to do when developing it but not when depending on it? (Or vice-versa?)

**Comment by zkat** [comment]
@janl you can use `$PWD` and see if it's inside `node_modules`.

**Comment by matthew-dean** [comment]
When I do `npm install -g npm@5` or `npm install -g npm@next` it runs the installer, but `npm -v` still shows NPM's version at `4.6.1`.

**Comment by zkat** [comment]
@matthew-dean what does `npm prefix -g` say? Is the path given there when you do `echo $PATH`?

**Comment by elyscape** [comment]
npm@5.0.0 is still marked as pre-release, [both on GitHub][github] and on npm itself:
```shellsession
scapeless:~ elyscape$ curl -s https://registry.npmjs.org/npm/ | jq '.["dist-tags"]'
{
  "latest": "4.6.1",
  "next": "5.0.0",
  "latest-2": "2.15.12",
  "next-2": "2.15.12",
  "latest-1": "1.4.29",
  "lts": "2.15.11",
  "latest-3": "3.10.10",
  "next-3": "3.10.10",
  "latest-4": "4.6.1",
  "next-4": "4.6.1"
}
```
[github]: https://github.com/npm/npm/releases/tag/v5.0.0

**Comment by SimenB** [comment]
(an easier command for ^ is just `npm dist-tags ls npm`)

**Comment by zkat** [comment]
Just to update here, it's `latest` now :)

**Comment by SimenB** [comment]
Nice! This issue can be closed, then?
Also, should a `{latest,next}-5` be created?

**Comment by zkat** [comment]
@SimenB done

**Comment by zkat** [comment]
I guess I should close this issue. :s I've been using it to track stuff but we're honestly in full live mode now, so... woo.

**Comment by janl** [comment]
> @janl you can use $PWD and see if it's inside node_modules.

@zkat that does the trick thank you!

**Comment by zkat** [comment]
p.s. there is now https://github.com/npm/npm/issues/16991 for post-release issue tracking
